### PR TITLE
New oidc fixes

### PR DESCRIPTION
**Pull request by mrvanes**
As requested on the idpy meeting, I hereby created a new pr containing separate commits for each sub-fix, a config option for oidc multi_value and an accompanying test for oidc multi_value. I hope it lives up to your expectations. There is also a whitespace fix, because my editor is set to remove trailing whitespace stuff.

### CODE REVIEW

File: tests/flows/test_oidc-saml.py [hunk 1]
```diff
@@ -98,3 +120,57 @@ def test_full_flow(self, satosa_config_dict, oidc_frontend_config, saml_backend_
...
+        authn_resp_req = urlparse(url).path + "?" + urlencode(authn_resp)
+        authn_resp = test_client.get("/" + authn_resp_req)
+        assert authn_resp.status == "303 See Other"
+
+        # verify auth resp from proxy
+        resp_dict = dict(parse_qsl(urlparse(authn_resp.data.decode("utf-8")).fragment))
+        signing_key = RSAKey(key=rsa_load(oidc_multi_value_frontend_config["config"]["signing_key_path"]),
+                             use="sig", alg="RS256")
+        id_token_claims = JWS().verify_compact(resp_dict["id_token"], keys=[signing_key])
+        assert all((k, v) in {ck:cv if isinstance(cv,list) else [cv] for ck, cv in id_token_claims.items()}.items() for k, v in USERS[user_id].items())
```
Comment (c00kiemon5ter) [inline tests/flows/test_oidc-saml.py]: This is a bit complex. I'm adding whitespace to make it readable

```python
assert all(
  (k, v) in {
    ck: cv if isinstance(cv,list) else [cv]
    for ck, cv in id_token_claims.items()
  }.items()
  for k, v in USERS[user_id].items())
```

What you're trying to do is figuring out if _all items_ from `USERS` match an item in `id_token_claims` after a transformation. This means that `USERS` is a subset of (the transformed) `id_token_claims`.

```python
claims_dict =  {
    ck: cv if isinstance(cv,list) else [cv]
    for ck, cv in id_token_claims.items()
}
claims_set = set(claims_dict.items())
users_set = set(USERS[user_id].items())
assert users_set.issubset(claims_set)
```

I think, though, that what you're really after is equality of the dicts.

```python
claims_dict =  {
    ck: cv if isinstance(cv,list) else [cv]
    for ck, cv in id_token_claims.items()
}
assert claims_dict == USERS[user_id]
```

File: src/satosa/frontends/openid_connect.py [hunk 2]
```diff
@@ -256,7 +256,13 @@ def _get_approved_attributes(self, provider_supported_claims, authn_req):
         if "claims" in authn_req:
             for k in ["id_token", "userinfo"]:
                 if k in authn_req["claims"]:
-                    requested_claims.extend(authn_req["claims"][k].keys())
+                    for sk, sv in authn_req["claims"][k].items():
+                        try:
```
[the remark this answers is not in the document]
Comment (mrvanes) [reply src/satosa/frontends/openid_connect.py]: Although I haven't tested this it sure looks to do the trick if "for" won't iterate over an empty list?

[the remark this answers is not in the document]
Comment (c00kiemon5ter) [reply src/satosa/frontends/openid_connect.py]: Yes, if the list is empty `for` will invoke the `else` block.

The two blocks below are equivalent:

```python
if lst:
  for item in lst:
    do_something_with(item)
else:
  do_something_with_empty_list()
```

```python
for item in lst:
  do_something_with(item)
else:
  do_something_with_empty_list()
```

[the remark this answers is not in the document]
Comment (mrvanes) [reply src/satosa/frontends/openid_connect.py]: Funny construct: for... else. Learning every day ;)

File: src/satosa/frontends/openid_connect.py [hunk 3]
```diff
@@ -256,7 +256,13 @@ def _get_approved_attributes(self, provider_supported_claims, authn_req):
         if "claims" in authn_req:
             for k in ["id_token", "userinfo"]:
                 if k in authn_req["claims"]:
-                    requested_claims.extend(authn_req["claims"][k].keys())
```
Comment (c00kiemon5ter) [inline src/satosa/frontends/openid_connect.py]: @mrvanes I'm going through this PR, but I'm sure how this is useful. This is some test data for `authn_req["claims"]`:

```python
authn_req["claims"] = {
  "id_token": {
    "foo": "bar",
    "mail": {
      "essential": True
    }
  },
  "userinfo": {
    "acr": {
      "essential": True,
      "values": [
        "urn:mace:incommon:iap:silver",
        "urn:mace:incommon:iap:bronze"
      ]
    }
  }
}
```

Coming out of the process, the result is:
```python
> print(requested_claims)
['sub', 'foo', 'email.essential', 'acr.essential', 'acr.values']
```

Do we really want `email.essential`? How is it more correct than before?

**Pull request closed without merging**